File: JS/Stack/4949.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let input = fs.readFileSync(filePath).toString().split("\n");

for (line of input) {
  if (line === ".") break;
  if (line) console.log(check(line) ? "yes" : "no");
}

function check(line) {
  let stack = new Array(101).fill();
  let p = -1;
  for (char of line) {
    if (char === "(" || char === "[") {
      stack[++p] = char;
    } else if (char === ")") {
      if (p === -1 || stack[p] !== "(") {
        return false;
      } else if (stack[p] === "(") {
        stack[p--] = 0;
      }
    } else if (char === "]") {
      if (p === -1 || stack[p] !== "[") {
        return false;
      } else if (stack[p] === "[") {
        stack[p--] = 0;
      }
    }
  }
  if (p === -1) return true;
  else return false;
}

File: Python/Stack_Queue/4949.py
import sys
input = sys.stdin.readline

def check (line):
  stack = []
  for c in line:
    if c == "[" or c == "(":
      stack.append(c)
    if c == "]":
      if not stack or stack.pop() != "[":
        return "no"
    if c == ")":
      if not stack or stack.pop() != "(":
        return "no"
      
  if len(stack) > 0:
    return "no"
  return "yes"

# while (line := input().replace('\n', '')) != '.':
while (line := input()[:-1]) != '.':
  print(check(line))
